language/resource_types.md

```diff
@@ -0,0 +1,440 @@
+# Puppet Resource Types
+
+This document describes puppet's DSL for defining puppet resource types. Examples are taken from the `puppetlabs-aws` module and core puppet.
+
+The second part of the document describes issues that cause puppet resource types to leak across environments (lack of environment isolation).
+
+## Puppet::Type.newtype method
+
+Defines a new puppet resource type and registers the type using the specified symbolic name. For example, to define a resource type named `ec2_instance`: 
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  # definition goes here
+end
+```
+
+The `newtype` method creates a new class `Puppet::Type::Ec2_instance`, which subclasses `Puppet::Type`. The block passed to the `newtype` method defines methods and variables on the `Puppet::Type::Ec2_instance` class.
+
+## Puppet::Type::&lt;name&gt; methods
+
+Within the `Puppet::Type.newtype` block, the following DSL methods are often called, and are evaluated in the context of the  `Puppet::Type::<name>` class:
+
+### doc string
+
+Defines the documentation string for the type, which is output when running `puppet describe <type>`. The doc string is usually set via the `Puppet::Type` class variable, though there is also a `Puppet::Type.doc=` class method.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  @doc = 'A type representing an AWS VPC'
+end
+```
+
+### title patterns
+
+Provides a way for setting attributes from the title. For example, the `file` resource sets its `path` namevar based on its `title`, stripping trailing slashes in the process. So if `title` is `/foo/bar//`, then `path` will be `/foo/bar`.
+
+```ruby
+Puppet::Type.newtype(:file) do
+  def self.title_patterns
+    [ [ /^(.*?)\/*\Z/m, [ [ :path ] ] ] ]
+  end
+end
+```
+
+A type must define a `title_patterns` method when using composite namevars so that puppet knows how to decompose the `title` into its constituent namevars, e.g. see [java-ks](https://github.com/puppetlabs/puppetlabs-java_ks/blob/abac95473a505080f60b4d0118c82d3568063da0/lib/puppet/type/java_ks.rb#L150-L176), [websphere](https://github.com/puppetlabs/puppetlabs-websphere_application_server/blob/cd2dbc59c43030db36a81f9c1ed155d5fbe4e85c/lib/puppet/type/websphere_sdk.rb#L11-L28).
+
+### ensureable
+
+Creates an `ensure` property with acceptable values of `present` and `absent`, each of which invoke the provider's `create` and `destroy` methods, respectively:
```

Comment:
For what it's worth, the fact that the package ensure value can `ensure => held` is deeply problematic because you can't hold a package at a specific version.

Oh, and then there's that - the package type also accepts a version as an ensure value, which amounts to it accepting an arbitrary string.

This is bad and terrible and makes me quite unhappy. I would love if we gradually deprecated these things and moved to ensure values being `present` and `absent` and everything else being shoved into a property.